File: 2/2_cuatrimestre/PDOO/practicas/practica 4/deepSpaceRuby/NumericDamage.rb
require_relative 'NumericDamageToUI'
require_relative 'Damage'
require_relative 'WeaponType'

module Deepspace
    class NumericDamage < Damage

        attr_reader :nWeapons

        public_class_method def initialize(w,s)
            super(s)
            @nWeapons = w
        end
        
        def copy()
            return new(@nWeapons,@nShields)
        end

        def getUIversion
            return NumericDamageToUI.new(self)
        end 

        def adjust(w, s)
            
            ns_aux = 0

            if @nShields > 0
                if @nShields > s.size
                    ns_aux = s.size
                else
                    ns_aux = @nShields
                end
            end
        
            nw_aux = 0
            if @nWeapons >= 0
                if @nWeapons > w.size
                    nw_aux = w.size
                else
                    nw_aux = @nWeapons
                end
            end
                
            return NumericDamage.new(nw_aux,ns_aux)
            
        end

        def discardWeapon(w)

            if (@nWeapons > 0)
                @nWeapons -= 1
            end
        end

        def hasNoEffect
            return ((@nShields == 0) && (@nWeapons == 0))
        end

        def to_s
            return getUIversion.to_s
        end

    end
end

File: 2/2_cuatrimestre/PDOO/practicas/practica 4/deepSpace_4/src/deepspace/NumericDamage.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package deepspace;

import java.util.ArrayList;

/**
 *
 * @author manbolq
 */
public class NumericDamage extends Damage{
    private int nWeapons;
    
    
    /**************************************************************************/
    /**************************    CONSTRUCTOR   ******************************/
    /**************************************************************************/
    
    NumericDamage(int w, int s){
        super(s);
        nWeapons = w;
    }    
    
    /**************************************************************************/
    /****************************    METODOS   ********************************/
    /**************************************************************************/
    
    public int getNWeapons(){
        return nWeapons;
    }
    
    @Override
    NumericDamageToUI getUIversion(){
        return new NumericDamageToUI(this);
    }

    @Override
    public Damage copy() {
        return new NumericDamage(nWeapons, nShields);
    }

    @Override
    public Damage adjust(ArrayList<Weapon> w, ArrayList<ShieldBooster> s) {
        int ns_aux = 0;
        if (nShields > 0){
            if (nShields > s.size())
                ns_aux = s.size();
            else
                ns_aux = nShields;
        }
            
        
        int nw_aux = 0;
        if (nWeapons > 0){
            if (nWeapons > w.size())
                nw_aux = w.size();
            else
                nw_aux = nWeapons;
        }
        
        return new NumericDamage(nw_aux, ns_aux);
        
    }

    @Override
    public void discardWeapon(Weapon w) {
        if (nWeapons > 0)
            nWeapons--;
    }

    @Override
    public boolean hasNoEffect() {
        return ( (nShields == 0) && (nWeapons == 0));
    }

    @Override
    public String toString() {
        return getUIversion().toString();
    }
    
    
}
